File: backend/providers/coingecko.py
import os
import requests
import logging

logger = logging.getLogger("tokencare-coingecko")
REQUEST_TIMEOUT = 4.0

def fetch_coingecko_token(contract_address: str, platform_id: str) -> dict:
    """
    Fetches token data from CoinGecko Contract API
    """
    api_key = os.getenv("COINGECKO_API_KEY", "")
    headers = {}
    if api_key:
        headers["x-cg-demo-api-key"] = api_key

    try:
        url = f"https://api.coingecko.com/api/v3/coins/{platform_id}/contract/{contract_address.lower()}"
        resp = requests.get(url, headers=headers, timeout=REQUEST_TIMEOUT)
        if resp.status_code == 200:
            data = resp.json()
            market_data = data.get("market_data", {})
            image_data = data.get("image", {})

            price = float((market_data.get("current_price") or {}).get("usd") or 0.0)
            price_change = float(market_data.get("price_change_percentage_24h") or 0.0)
            market_cap = float((market_data.get("market_cap") or {}).get("usd") or 0.0)
            logo = image_data.get("large") or image_data.get("small") or None

            return {
                "source": "coingecko",
                "found": True,
                "name": data.get("name"),
                "symbol": str(data.get("symbol") or "").upper(),
                "logo": logo,
                "price": price,
                "priceUsd": price,
                "priceChange24h": price_change,
                "marketCap": market_cap,
                "volume24h": float((market_data.get("total_volume") or {}).get("usd") or 0.0),
                "verified": True
            }
    except Exception as e:
        logger.warning(f"[CoinGecko] Error fetching {contract_address}: {e}")

    return {"source": "coingecko", "found": False}

File: backend/providers/coingecko.ts
import { NormalizedTokenDetails } from '../types/tokenDetails';

export const CHAIN_MAP_COINGECKO: Record<string, string> = {
  '1': 'ethereum',
  '137': 'polygon-pos',
  '56': 'binance-smart-chain',
  '42161': 'arbitrum-one',
  '10': 'optimistic-ethereum',
  '8453': 'base',
  '43114': 'avalanche',
  'solana': 'solana',
  'ethereum': 'ethereum',
  'polygon': 'polygon-pos',
};

export async function fetchCoinGeckoToken(
  chain: string,
  contractAddress: string
): Promise<NormalizedTokenDetails | null> {
  const cleanAddr = contractAddress.trim().toLowerCase();
  if (!cleanAddr) return null;

  try {
    const platform = CHAIN_MAP_COINGECKO[chain.toLowerCase()] || 'ethereum';
    const url = `https://api.coingecko.com/api/v3/coins/${platform}/contract/${cleanAddr}`;

    const res = await fetch(url, {
      headers: { 'Accept': 'application/json' },
    });

    if (!res.ok) return null;
    const data = await res.json();
    if (!data || !data.name) return null;

    const marketData = data.market_data;
    const isSolana = platform === 'solana';

    const normalized: NormalizedTokenDetails = {
      name: data.name || 'Unknown Token',
      symbol: (data.symbol || 'TOKEN').toUpperCase(),
      decimals: isSolana ? 9 : 18,
      contractAddress: cleanAddr,
      chain: chain,
      chainId: chain,
      blockchain: isSolana ? 'Solana' : 'EVM',
      assetStandard: isSolana ? 'SPL' : 'ERC20',
      logoUrl: data.image?.large || data.image?.small || undefined,
      websiteUrl: data.links?.homepage?.[0] || undefined,
      twitterUrl: data.links?.twitter_screen_name ? `https://twitter.com/${data.links.twitter_screen_name}` : undefined,
      telegramUrl: data.links?.telegram_channel_identifier ? `https://t.me/${data.links.telegram_channel_identifier}` : undefined,
      priceUsd: marketData?.current_price?.usd || undefined,
      fdvUsd: marketData?.fully_diluted_valuation?.usd || undefined,
      marketCapUsd: marketData?.market_cap?.usd || marketData?.fully_diluted_valuation?.usd || undefined,
      volume24hUsd: marketData?.total_volume?.usd || 0,
      priceChange24h: marketData?.price_change_percentage_24h || 0,
      totalSupply: marketData?.total_supply ? String(marketData.total_supply) : undefined,
      resolvedVia: 'coingecko',
    };

    return normalized;
  } catch (err) {
    console.error('[CoinGecko Provider Error]:', err);
    return null;
  }
}
